service/frontend/redirection_interceptor.go

```diff
@@ -0,0 +1,254 @@
+
+		"StartBatchOperation":    func() interface{} { return new(workflowservice.StartBatchOperationResponse) },
+		"StopBatchOperation":     func() interface{} { return new(workflowservice.StopBatchOperationResponse) },
+		"DescribeBatchOperation": func() interface{} { return new(workflowservice.DescribeBatchOperationResponse) },
+		"ListBatchOperations":    func() interface{} { return new(workflowservice.ListBatchOperationsResponse) },
+	}
+)
+
+type (
+	respAllocFn func() interface{}
+
+	// RedirectionInterceptor is simple wrapper over frontend service, doing redirection based on policy
+	RedirectionInterceptor struct {
+		currentClusterName string
+		config             *Config
+		namespaceCache     namespace.Registry
+		redirectionPolicy  DCRedirectionPolicy
+		logger             log.Logger
+		clientBean         client.Bean
+		metricsHandler     metrics.Handler
+		timeSource         clock.TimeSource
+	}
+)
+
+// NewRedirectionInterceptor creates DC redirection interceptor
+func NewRedirectionInterceptor(
+	configuration *Config,
+	namespaceCache namespace.Registry,
+	policy config.DCRedirectionPolicy,
+	logger log.Logger,
+	clientBean client.Bean,
+	metricsHandler metrics.Handler,
+	timeSource clock.TimeSource,
+	clusterMetadata cluster.Metadata,
+) *RedirectionInterceptor {
+	dcRedirectionPolicy := RedirectionPolicyGenerator(
+		clusterMetadata,
+		configuration,
+		namespaceCache,
+		policy,
+	)
+
+	return &RedirectionInterceptor{
+		currentClusterName: clusterMetadata.GetCurrentClusterName(),
+		config:             configuration,
+		redirectionPolicy:  dcRedirectionPolicy,
+		namespaceCache:     namespaceCache,
+		logger:             logger,
+		clientBean:         clientBean,
+		metricsHandler:     metricsHandler,
+		timeSource:         timeSource,
+	}
+}
+
+var _ grpc.UnaryServerInterceptor = (*RedirectionInterceptor)(nil).Intercept
+
+func (i *RedirectionInterceptor) Intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ interface{}, retError error) {
+	defer log.CapturePanic(i.logger, &retError)
+
+	_, methodName := interceptor.SplitMethodName(info.FullMethod)
+	if _, ok := localAPIResults[methodName]; ok {
+		return i.handleLocalAPIInvocation(ctx, req, handler, methodName)
+	}
+	if respAllocFn, ok := globalAPIResults[methodName]; ok {
+		namespaceName := interceptor.GetNamespace(i.namespaceCache, req)
+		return i.handleRedirectAPIInvocation(ctx, req, info, handler, methodName, respAllocFn, namespaceName)
+	}
+
+	// this should not happen, log warn and move on
+	i.logger.Warn(fmt.Sprintf("RedirectionInterceptor encountered unknown API: %v", methodName))
+	return handler(ctx, req)
+}
+
+func (i *RedirectionInterceptor) handleLocalAPIInvocation(
+	ctx context.Context,
+	req interface{},
+	handler grpc.UnaryHandler,
+	methodName string,
+) (_ interface{}, retError error) {
+	scope, startTime := i.beforeCall(dcRedirectionMetricsPrefix + methodName)
```

Comment:
only method name if it is local invocation?

Reply:
what do you mean?